include/nbl/builtin/hlsl/emulated/float64_t.hlsl

```diff
@@ -0,0 +1,657 @@
+                    // can only happen when shiftAmount == 0, so it is safe to swap only high bits of rhs mantissa
+                    if (mantissaDiff < 0)
+                    {
+                        swap<uint64_t>(lhsNormMantissa, rhsNormMantissaHigh);
+                        swap<uint64_t>(lhsSign, rhsSign);
+                    }
+
+                    resultMantissa = emulated_float64_t_impl::subMantissas128NormalizeResult(lhsNormMantissa, rhsNormMantissaHigh, rhsNormMantissaLow, exp);
+
+                    if (resultMantissa == 0ull)
+                        return _static_cast<this_t>(0ull);
+                }
+                else
+                {
+                    rhsNormMantissa >>= shiftAmount;
+                    resultMantissa = lhsNormMantissa + rhsNormMantissa;
+
+                    if (resultMantissa & 1ull << 53)
+                    {
+                        ++exp;
+                        resultMantissa >>= 1;
+                    }
+                }
+
+                uint64_t resultBiasedExp = uint64_t(exp) + ieee754::traits<float64_t>::exponentBias;
+                resultMantissa &= ieee754::traits<float64_t>::mantissaMask;
+                uint64_t output = emulated_float64_t_impl::assembleFloat64(lhsSign, resultBiasedExp << ieee754::traits<float64_t>::mantissaBitCnt, resultMantissa);
+                return bit_cast<this_t>(output);
+            }
+
+            // not implemented
+            if (!FlushDenormToZero)
+                return bit_cast<this_t>(0xdeadbeefbadcaffeull);
+        }
+
+        emulated_float64_t operator+(float rhs)
+        {
+            return bit_cast<this_t>(data) + create(rhs);
+        }
+
+        emulated_float64_t operator-(emulated_float64_t rhs) NBL_CONST_MEMBER_FUNC
+        {
+            emulated_float64_t lhs = bit_cast<this_t>(data);
+            emulated_float64_t rhsFlipped = rhs.flipSign();
+            
+            return lhs + rhsFlipped;
+        }
+
+        emulated_float64_t operator-(float rhs) NBL_CONST_MEMBER_FUNC
+        {
+            return bit_cast<this_t>(data) - create(rhs);
+        }
+
+        emulated_float64_t operator*(emulated_float64_t rhs) NBL_CONST_MEMBER_FUNC
+        {
+            // TODO: remove
+            float64_t sum = bit_cast<float64_t>(data) * bit_cast<float64_t>(rhs.data);
+            uint64_t sumAsUint = bit_cast<uint64_t>(sum);
+
+            this_t output2;
+            output2.data = sumAsUint;
+
+            return output2;
+
+            if(FlushDenormToZero)
+            {
+                emulated_float64_t retval = this_t::create(0ull);
+
+                int lhsBiasedExp = ieee754::extractBiasedExponent(data);
+                int rhsBiasedExp = ieee754::extractBiasedExponent(rhs.data);
+
+                uint64_t lhsData = emulated_float64_t_impl::flushDenormToZero(lhsBiasedExp, data);
+                uint64_t rhsData = emulated_float64_t_impl::flushDenormToZero(rhsBiasedExp, rhs.data);
```

Comment:
why do you produce new data?, you could just check for them being denormalized and return a signed 0 right away

Reply:
actually let me flip this on its head, if you flush to 0 the OUTPUT of every operation, there's no need to check whether the inputs need to be flushed to 0 ever.